docs/integration_fetching_data.md

```diff
@@ -0,0 +1,146 @@
+---
+title: "Fetching Data"
+---
+
+Your integration will need to fetch data from an API to be able to provide this to Home Assistant. This API can be available over the web (local or cloud), sockets, serial ports exposed via USB sticks, etc.
+
+# Push vs Poll
+
+APIs come in many different shapes and forms but at its core they fall in two categories: push and poll.
+
+With push, we subscribe to an API and we get notified by the API when new data is available. It pushes the changes to us. Push APIs are great because they consume less resources. When a change happens, we can get notified of a change and don't have to re-fetch all the data and find changes. Because entities can be disabled, you should make sure that your entity subscribes inside the `async_added_to_hass` callback and unsubscribes inside `async_will_remove_from_hass`.
+
+With polling, we will fetch the latest data from the API at a specified interval. Your integration will then supply this data to its entity, which is written to Home Assistant.
+
+Because polling is so common, Home Assistant by default assumes that your entity is based on polling. If this is not the case, return `False` from the `should_poll` property. When you disable polling, your integration will be responsible for calling one of the methods to indicate to Home Assistant that it's time to write the entity state to Home Assistant:
+
+- `Entity.schedule_update_ha_state(force_refresh=False)`/`Entity.async_schedule_update_ha_state(force_refresh=False)` will schedule an update of the entity. If `force_refresh` is set to `True`, Home Assistant will call your entities update method (`update()`/`async_update()`) prior to writing the state.
+- `Entity.async_write_ha_state()` is an async callback that will write the state to the state machine within yielding to the event loop. It will not call your entity update method.
+
+# Polling API endpoints
+
+We're going to explain a few different API types here and the best way to integrate them in Home Assistant. Note that some integrations will encounter a combination of the ones below.
+
+## Coordinated, single API poll for data for all entities
+
+This API will have a single method to fetch data for all the entities that you have in Home Assistant. In this case we will want to have a single periodical poll on this endpoint, and then let entities know as soon as new data is available for them.
+
+Home Assistant provides a DataUpdateCoordinator class to help you manage this as efficiently as possible.
+
+```python
+import logging
+
+from homeassistant.helpers import debounce, entity
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+
+_LOGGER = logging.getLogger(__name__)
+
+async def create_coordinator(api):
+    async def async_update_data():
+        """Fetch data from API endpoint."""
+        try:
+            await api.fetch_data()
+        except ApiError:
+            raise UpdateFailed
```

Comment:
API-wise, IMO, it would be much nicer to simply let the update method to raise any exception to be handled accordingly by the coordinator? This would simplify implementations by not forcing developers to remember to do the import and do a try-catch for their update code.

Reply:
We want to be able to distinguish between expected and unexpected errors. Unexpected errors will be logged with a stack trace.